src/log.rs

```diff
@@ -81,13 +82,316 @@ type CompatFilterStream<T, R> = Compat01As03<FilterStream<T, R>>;
+            filter = filter.from_block(from_block);
+        }
+        if let Some(to_block) = self.to_block {
+            filter = filter.to_block(to_block);
+        }
+
+        filter
+    }
+}
+
+impl<'t, T> LogFilterBuilder<T>
+where
+    T: Transport + Send + Sync + 't,
+    T::Out: Send,
+{
+    /// Returns a stream that resolves into a page of logs matching the filter
+    /// builder's parameters.
+    ///
+    /// Note this is a failure tolerant stream in that when a failure is
+    /// encountered and `Some(Err(_))` is yielded, polling the stream again will
+    /// resume from the last error and will not return `None`.
+    pub fn past_logs(self) -> BoxStream<'t, Result<Vec<Log>, ExecutionError>> {
+        stream::unfold(PastLogsPager::from_builder(self), |mut pager| async move {
+            let page = pager.next_page().await.transpose()?;
+            Some((page, pager))
+        })
+        .boxed()
+    }
+}
+
+/// Internal state for paging though past logs.
+struct PastLogsPager<T: Transport> {
+    web3: Web3<T>,
+
+    /// The `to_block` specified by the log filter.
+    to_block: BlockNumber,
+    /// The block page size being used for queries.
+    block_page_size: u64,
+    /// The web3 filter used for retrieving the logs.
+    filter: FilterBuilder,
+
+    /// The block number for the next page.
+    ///
+    /// This value can be `None` if it `from_block` was specified as `Latest` or
+    /// `Pending`.
+    page_block: Option<u64>,
+    /// The last block used for pagination. This is slightly different than
+    /// `to_block` as this must be a concrete block number (and can't be block
+    /// aliases such as `Earliest` or `Latest`).
+    ///
+    /// This has a value of `None` if it has not yet been retrieved (in case
+    /// `to_block` is `Latest` or `Pending`).
+    end_block: Option<u64>,
+}
+
+impl<T: Transport> PastLogsPager<T> {
+    fn from_builder(builder: LogFilterBuilder<T>) -> Self {
+        let web3 = builder.web3.clone();
+        let to_block = builder.to_block.unwrap_or(BlockNumber::Latest);
+        let block_page_size = builder.block_page_size.unwrap_or(DEFAULT_BLOCK_PAGE_SIZE);
+
+        let page_block = match builder.from_block.unwrap_or(BlockNumber::Latest) {
+            BlockNumber::Earliest => Some(0),
+            BlockNumber::Number(value) => Some(value.as_u64()),
+            BlockNumber::Latest | BlockNumber::Pending => None,
+        };
+        let end_block = match to_block {
+            BlockNumber::Earliest => Some(0),
+            BlockNumber::Number(value) => Some(value.as_u64()),
+            BlockNumber::Latest | BlockNumber::Pending => None,
+        };
+
+        let filter = builder.into_filter();
+
+        PastLogsPager {
+            web3,
+            to_block,
+            block_page_size,
+            filter,
+            page_block,
+            end_block,
+        }
+    }
+
+    async fn next_page(&mut self) -> Result<Option<Vec<Log>>, ExecutionError> {
```

Comment:
Split into multiple smaller functions if sensible. If it doesn't turn out to work well then not needed.

Reply:
I think I need to refactor this a bit. You are right, things are a little complicated at the moment.